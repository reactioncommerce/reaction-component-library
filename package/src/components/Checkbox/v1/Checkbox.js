import React, { Component } from "react";
import PropTypes from "prop-types";
import uniqueId from "lodash.uniqueid";
import styled from "styled-components";
import { applyTheme } from "../../../utils";

const StyledDiv = styled.div`
  margin-bottom: ${applyTheme("checkboxVerticalSpacing")};
`;

const StyledInput = styled.input`
  opacity: 0;
  + label::after {
    content: none;
  }
  &:checked + label::after {
    content: "\f00c";
  }
  &:focus + label::before {
    outline: ${applyTheme("checkboxFocusStyle")};
  }
  &:disabled + label {
    opacity: ${applyTheme("checkboxDisabledOpacity")};
  }
  &:disabled + label::before {
    background: ${applyTheme("checkboxDisabledColor")};
    opacity: ${applyTheme("checkboxDisabledOpacity")};
  }
  &:disabled + label:hover {
    cursor: not-allowed;
  }
`;

const StyledLabel = styled.label`
  position: relative;
  display: inline-block;
  font-size: ${applyTheme("checkboxLabelFontSize")};
  padding-left: ${applyTheme("checkboxLabelSpacing")};
  &:hover {
    cursor: pointer;
  }
  &::before,
  &::after {
    position: absolute;
  }
  &::before {
    content: " ";
    display: inline-block;
    height: ${applyTheme("checkboxHeightAndWidth")};
    width: ${applyTheme("checkboxHeightAndWidth")};
    border: ${applyTheme("checkboxBorderWidth")} solid ${applyTheme("checkboxBorderColor")};
    border-radius: ${applyTheme("checkboxBorderRadius")};
    left: 0;
    top: -3px;
  }
  &::after {
    content: "\f00c";
    display: inline-block;
    font-family: FontAwesome;
    font-size: ${applyTheme("checkboxIconSize")};
    color: ${applyTheme("checkboxIconColor")};
    left: 5px;
    top: 2px;
  }
`;

class Checkbox extends Component {
  static propTypes = {
    className: PropTypes.string,
    isReadOnly: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
    label: PropTypes.string.isRequired,
    name: PropTypes.string,
    onChange: PropTypes.func,
    onChanging: PropTypes.func,
    value: PropTypes.bool
  };

  static defaultProps = {
    className: undefined,
    isReadOnly: false,
    name: undefined,
    onChange() { },
    onChanging() { },
    value: undefined
  };

  constructor(props) {
    super(props);

    this.state = {
      id: uniqueId("Checkbox_"),
      value: props.value || false
    };
  }

  componentWillMount() {
    this.handleChange(this.props.value || false);
  }

  componentWillReceiveProps(nextProps) {
    const { value } = this.props;
    const { value: nextValue } = nextProps;

    // Whenever a changed value prop comes in, we reset state to that, thus becoming clean.
    if (value !== nextValue) {
      this.setState({ value: nextValue || false });
      this.handleChange(nextValue || false);
    }
  }

  onChange = (event) => {
    this.setValue(event.target.checked);
  };

  getValue() {
    return this.state.value;
  }

  setValue(value) {
    this.setState({ value });
    this.handleChange(value);
  }

  resetValue() {
    this.setValue(this.props.value || false);
  }

  handleChange(checked) {
    if (this.lastValue === checked) return;
    this.lastValue = checked;
    const { onChanging, onChange } = this.props;
    onChanging(checked);
    onChange(checked);
  }

  // Input is dirty if value prop doesn't match value state. Whenever a changed
  // value prop comes in, we reset state to that, thus becoming clean.
  isDirty() {
    return this.state.value !== this.props.value;
  }

  render() {
    const { className, isReadOnly, label } = this.props;
    const { id, value } = this.state;

    return (
      <StyledDiv className={className}>
        <StyledInput
          checked={value === true}
          id={id}
          onChange={this.onChange}
          disabled={isReadOnly}
          type="checkbox"
          value="true"
        />
        <StyledLabel htmlFor={id}>
          {label}
        </StyledLabel>
      </StyledDiv>
    );
  }
}

export default Checkbox;
