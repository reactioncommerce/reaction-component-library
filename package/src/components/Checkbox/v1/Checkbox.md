Checkboxes can be checked, unchecked and disabled.

```jsx
<Field name="example">
  <Checkbox name="example-checked" label="Checked" value={true} />
  <Checkbox name="example-unchecked" label="Unchecked" />
  <Checkbox name="example-disabled" label="Disabled" isReadOnly />
</Field>
```

#### Usage

###### Checkbox

- Checkboxes should be used when more than one option can be enabled.
- Generally, checkboxes should be used when there are 3 or more options that can be selected in a row.
- Checkboxes should be left-aligned in forms.

##### Label

- All labels should be written in sentence caps.
- Labels should be left-aligned in forms.

#### Specs

##### Checkbox

The checkbox is a styled `::before` pseudoelement on `<input>`. The native `<input>` is hidden with an `opacity` set to `0`.

|Property     |Style               |
|-------------|:------------------:|
|Cursor       |Pointer             |
|Height       |`@base-unit(2)`     |
|Width        |`@base-unit(2)`     |
|Border       |`2px @cool-grey-500`|
|Border radius|`2px @cool-grey-500`|

##### Checkbox check

The checkbox's check icon an `::after` pseudoelement on `<input>`, with a `content` set to FontAwesome's [check](https://fontawesome.com/icons/check?style=solid): <i class="fas fa-check"></i>

|Property     |Style               |
|-------------|:------------------:|
|Color        |`@cool-grey-500`    |
|Font family  |`FontAwesome`       |
|Font size    |`0.875rem`          |
|Content      |`\f00c`             |

##### Label

|Property     |Style               |
|-------------|:------------------:|
|Cursor       |Pointer             |
|Font size    |`@rui-label-text`   |
|Color        |`@cool-grey-500    `|

##### Disabled

|Property      |Style               |
|--------------|:------------------:|
|Cursor        |Not allowed         |
|Checkbox fill |`@black10`          |
|Opacity       |`50%`               |

##### Structure

|Property                   |Spacing        |
|---------------------------|:-------------:|
|Between checkbox and label |`@base-unit(1)`|
|Vertical                   |`@base-unit(2)`|