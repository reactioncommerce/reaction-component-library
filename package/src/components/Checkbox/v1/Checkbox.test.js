import React from "react";
import renderer from "react-test-renderer";
import Checkbox from "./Checkbox";

test("renders with props", () => {
  const component = renderer.create(<Checkbox label="Label" />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders disabled", () => {
  const component = renderer.create(<Checkbox label="Disabled" isReadOnly />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
