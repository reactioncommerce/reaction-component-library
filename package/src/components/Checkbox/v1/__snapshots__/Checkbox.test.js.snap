// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with props 1`] = `
.c0 {
  margin-bottom: 1.25rem;
}

.c1 {
  opacity: 0;
}

.c1 + label::after {
  content: none;
}

.c1:checked + label::after {
  content: "\\f00c";
}

.c1:focus + label::before {
  outline: rgb(59,153,252) auto 5px;
}

.c1:disabled + label {
  opacity: .5;
}

.c1:disabled + label::before {
  background: #e6e6e6;
  opacity: .5;
}

.c1:disabled + label:hover {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  display: inline-block;
  font-size: 14px;
  padding-left: 1.875rem;
}

.c2:hover {
  cursor: pointer;
}

.c2::before,
.c2::after {
  position: absolute;
}

.c2::before {
  content: " ";
  display: inline-block;
  height: 1.25rem;
  width: 1.25rem;
  border: 2px solid #3c3c3c;
  border-radius: 2px;
  left: 0;
  top: -3px;
}

.c2::after {
  content: "\\f00c";
  display: inline-block;
  font-family: FontAwesome;
  font-size: 0.875rem;
  color: #3c3c3c;
  left: 5px;
  top: 2px;
}

<div
  className="c0"
>
  <input
    checked={false}
    className="c1"
    disabled={false}
    id="Checkbox_1"
    onChange={[Function]}
    type="checkbox"
    value="true"
  />
  <label
    className="c2"
    htmlFor="Checkbox_1"
  >
    Label
  </label>
</div>
`;

exports[`renders disabled 1`] = `
.c0 {
  margin-bottom: 1.25rem;
}

.c1 {
  opacity: 0;
}

.c1 + label::after {
  content: none;
}

.c1:checked + label::after {
  content: "\\f00c";
}

.c1:focus + label::before {
  outline: rgb(59,153,252) auto 5px;
}

.c1:disabled + label {
  opacity: .5;
}

.c1:disabled + label::before {
  background: #e6e6e6;
  opacity: .5;
}

.c1:disabled + label:hover {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  display: inline-block;
  font-size: 14px;
  padding-left: 1.875rem;
}

.c2:hover {
  cursor: pointer;
}

.c2::before,
.c2::after {
  position: absolute;
}

.c2::before {
  content: " ";
  display: inline-block;
  height: 1.25rem;
  width: 1.25rem;
  border: 2px solid #3c3c3c;
  border-radius: 2px;
  left: 0;
  top: -3px;
}

.c2::after {
  content: "\\f00c";
  display: inline-block;
  font-family: FontAwesome;
  font-size: 0.875rem;
  color: #3c3c3c;
  left: 5px;
  top: 2px;
}

<div
  className="c0"
>
  <input
    checked={false}
    className="c1"
    disabled={true}
    id="Checkbox_2"
    onChange={[Function]}
    type="checkbox"
    value="true"
  />
  <label
    className="c2"
    htmlFor="Checkbox_2"
  >
    Disabled
  </label>
</div>
`;
